File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    
    ListNode* solve(ListNode* l1, ListNode* l2, int carry){
        if(l1 == NULL && l2 == NULL && carry == 0){
            return NULL;
        }
        
        int sum = carry;
        if(l1!=NULL){
            sum = sum + l1->val;
        }else{
            sum = sum + 0;
        }
        
        if(l2!=NULL){
            sum = sum + l2->val;
        }else{
            sum = sum + 0;
        }
        
        ListNode*newNode=new ListNode(sum%10);
        carry=sum/10;
        
        if(l1) l1=l1->next;
        if(l2) l2=l2->next;
        newNode->next=solve(l1,l2,carry);
        return newNode;
        
    }
    
    
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        return solve(l1, l2, 0);
    }
};

File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode ans = new ListNode(-1);
        ListNode res = ans;
        ListNode temp1 = l1;
        ListNode temp2 = l2;
        int carry = 0;

        while (temp1 != null || temp2 != null || carry != 0) {
            int sum = carry;
            if (temp1 != null) {
                sum += temp1.val;
                temp1 = temp1.next;
            }
            if (temp2 != null) {
                sum += temp2.val;
                temp2 = temp2.next;
            }
            carry = sum / 10;
            res.next = new ListNode(sum % 10);
            res = res.next;
        }
        return ans.next;
    }
}
